packages/debugger/src/panels/kernelSources/body.tsx

```diff
@@ -21,19 +23,146 @@ import { UseSignal } from '@jupyterlab/ui-components';
 import { IRenderMime } from '@jupyterlab/rendermime';
 
 /**
- * The class name added to the filterbox node.
+ * CSS class names.
  */
 const FILTERBOX_CLASS = 'jp-DebuggerKernelSource-filterBox';
+const FILTERBOX_HIDDEN_CLASS = 'jp-DebuggerKernelSource-filterBox-hidden';
+const SOURCE_CLASS = 'jp-DebuggerKernelSource-source';
+const DIR_CLASS = 'jp-DebuggerKernelSource-dir';
 
 /**
- * The class name added to hide the filterbox node.
+ * Props for a tree node.
  */
-const FILTERBOX_HIDDEN_CLASS = 'jp-DebuggerKernelSource-filterBox-hidden';
+interface ITreeNodeProps {
+  name: string;
+  path: string;
+  children?: ITreeNodeProps[];
+  onOpen: (path: string) => void;
+}
 
 /**
- * The class for each source row.
+ * Recursive component for rendering a tree node (directory or file).
  */
-const SOURCE_CLASS = 'jp-DebuggerKernelSource-source';
+function TreeNode(props: ITreeNodeProps): JSX.Element {
+  const { name, path, children, onOpen } = props;
+  const [isOpen, setIsOpen] = useState(false);
+
+  const isDirectory = !!children && children.length > 0;
+
+  // Compute collapsed chain dynamically (only when expanded)
+  let displayName = name;
+  let displayChildren = children;
+
+  if (isOpen && isDirectory) {
+    const chain: string[] = [name];
+    let currentChildren = children;
+
+    // collapse linear folder chains dynamically
+    while (
+      currentChildren &&
+      currentChildren.length === 1 &&
+      currentChildren[0].children &&
+      currentChildren[0].children!.length > 0
+    ) {
+      const next = currentChildren[0];
+      chain.push(next.name);
+      currentChildren = next.children ?? [];
+    }
+
+    if (chain.length > 1) {
+      const sep = path.includes('.') ? '.' : '/';
```

Comment:
This condition will always be true since there will most always be a `.` in the path? e.g. both `include/folder/module.cpp` and `site-packages/folder/module.py` contain dots but for the cpp files the separators will most likely be `/`